Updated exclude_border to not use min_distance

jmetz opened the pull request:
Addresses #1950 

`min_distance` now not used, in favour of exclude_border specifying the border to exclude.

soupault commented on skimage/feature/peak.py:
```diff
@@ -137,12 +136,12 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
             mask = mask.swapaxes(0, i)
             remove = (footprint.shape[i] if footprint is not None
-                      else 2 * min_distance)
+                      else exclude_border)
```
What if `exclude_border==-2`? This will completely destroy the image.

soupault replied:
Moreover, it should be twice as much ( see division by 2 below ).

jmetz commented:
@soupault - the same is true of `min_distance` (AFAIK) - so unless we explicitly check all such inputs, checking this one makes no sense, or did I miss something?

soupault commented on skimage/feature/peak.py:
```diff
@@ -24,17 +24,16 @@ def peak_local_max(image, min_distance=1, threshold_abs=None,
-        well as from each other.
+    exclude_border : int, optional
+        If nonzero, `exclude_border` excludes peaks from
+        within `exclude_border` of the border of the image.
```
Could you reformulate this?
Like, for example, `peaks within 'exclude_border'-pixels frame will be excluded`.

soupault commented:
@jmetz you are right. 
We actually have several additional issues with the `peak` functions ( see https://github.com/scikit-image/scikit-image/pull/1932, for example ).
So, we'll be very grateful if you could contribute some extra time partiticipating in the discussion or even improving the code itself.

JDWarner commented:
While this clearly seemed unintuitive behavior to you (as described in #1950), these changes break backwards compatibility. Therefore, changing this requires either:

* A _bug_, i.e., incorrect/improper behavior, which justifies a breaking API change
* Deprecation path across a minimum of two point releases, with current behavior maintained in the interim

As the prior behavior did indeed work as it was documented (though not as was expected), it's hard to make the former case for an immediate, breaking change.

jni commented:
@jmetz @JDWarner yeah I think the backwards compatibility needs to be preserved, but perhaps we don't even need a deprecation path:

```python
if type(exclude_border) == bool:
    exclude_border = min_distance if exclude_border else 0
```

jmetz commented:
Thanks @jni - copied in the backwards compatibility patch.

jni commented:
@jmetz it looks like the change of default value made a whole bunch of tests fail. I think we should leave the default as True, again for backwards compatibility. If you think the default should be 1 in the future, then we should follow the usual deprecation path:

- default to None
- check if None, then make it True, but raise a warning that it will default to 1 in 0.14
- add "change peak_local_max exclude_borders default to 1" to TODO.txt under 0.14

Whether to include this change is entirely up to you. =)

The failing tests reminded me though that you should make a test for this new functionality...

ahojnnes commented:
Why default to 1 and not to 0?

jni commented:
@ahojnnes could be 0, I was merely translating the current default of True.

jmetz commented:
@jni - if I change `exclude_border` to default to `True`, this should fix the backwards-compatibility issues (as it'll get set to `min_distance`). 

Do you think that's sufficient?

jni commented:
@jmetz yes it is, but you should still add a test for the behaviour when an int is passed.

jni commented:
@jmetz all code paths should be tested, so we don't want to change *all* values to ints. Rather, you should make sure that at least one each is run of exclude border = True, False, 0, nonzero-int.

jmetz commented:
@jni - I think I spotted the issue; I reverted tests to using bools as requested, but the actual issue was still checking for footprint, i.e. 
`... and (footprint is not None)`
in line 142. 
I removed this as it was only needed for the previous use of `min_distance`, now all tests passed (at least on my machine!).

jni commented:
Great! :+1: to merge from me!

ahojnnes commented:
Great, thanks @jmetz !

jmetz commented:
Np - thanks for the friendly reminders, a bit rusty wrt skimage contribution - aim to be more active though. 

Speaking of which, what's the pathway for getting write access?

jni commented:
@jmetz there's been some recent talk of adopting a formal process, but until now it's been something like, "make a bunch of 'major' PRs in a row until @stefanv notices". Sorry it's not more concrete yet! =)

stefanv commented:
That said, write access is much less important nowadays with pull requests.
We all go through the same procedure anyway!

jni commented:
@stefanv @jmetz indeed, good point to make. All code, including that from core developers, must be approved by two other core developers before getting merged. So really it's more of a hassle than anything else. =P

jmetz commented:
Thanks @stefanv & @jni - very good points, cheers!
